### PR TITLE
Implement auth APIs

### LINKED ISSUE #29: Implement basic auth APIs
- `/auth?action=signup`
- `/auth?action=signin`
- `/auth?action=signout`

### PR DESCRIPTION
**Pull request by lvnam96**
ImplementedResolves auth APIs:
- Sign up (password)
- Sign in (password)
- Sign out
- Refresh tokens (extend session)

Resolves #29.

## Summary by CodeRabbit

- **New Features**
  - Enhanced authentication workflows (signup, signin, signout) and dynamic URL management.
  - Expanded configuration options with new environment variables.
  - Introduced a caching mechanism for HTTP request data.
  - Added utility functions for URL query parameter handling.

- **Documentation**
  - Updated project documentation with new badges, Docker container details, and guidance on environment variables.

- **Tests**
  - Broadened test coverage across middleware, API responses, JWT, password handling, and various utility functions.

- **Chores**
  - Improved database connectivity, error handling, and request caching alongside updated dependencies and build configurations.

### CODE REVIEW

File: app/.server/modules/auth/index.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  const { passwd, salt } = await getSaltedPassword(user.password);
+
+  return createUser({
+    ...user,
+    displayedName: user.displayedName || user.email,
+    salt,
+    password: await hashPassword(passwd),
+  });
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: How would I add DB constraint or check for a signup is in progress?

File: app/.server/utils/api.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  errors: ApiResponseError['errors'],
+  meta: ApiResponseError['meta'] = null,
+): ApiResponseError => ({
+  data: null,
+  errors,
+  meta,
+  links: null,
+});
+
+export const getBearerTokenFromAuthHeader = (authHeader: string) => authHeader.substring('Bearer '.length) || null;
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/utils/api.ts]: AI suggested code is too much, this code has been updated to this instead:
```ts
export const getBearerTokenFromAuthHeader = (authHeader: string) => authHeader?.substring?.('Bearer '.length) || null;
```

File: app/.server/api/auth/index.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        ),
+      );
+    } catch (err) {
+      if (err instanceof Response) throw err;
+      console.error(err);
+      return getUnauthorizedResponse();
+    }
+  } else if (request.method === 'POST' && action === 'signout') {
+    try {
+      const { tokenPayload } = await getRequestData<RefreshTokenPayload>(actionArgs);
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/auth/index.ts]: this is wrong due to missing whole repo's context

File: app/.server/utils/jwt.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    algorithm = 'HS512',
+  }: {
+    secret?: Parameters<SignJWT['sign']>[0];
+    expirationTime?: Parameters<ProduceJWT['setExpirationTime']>[0];
+    algorithm?: JWTHeaderParameters['alg'];
+  } = {
+    expirationTime: '1w',
+    algorithm: 'HS512',
+  },
+) => {
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/utils/jwt.ts:32]: This is wrong since any fields can be missing so default value must be set for each one of them. If entire option object is missing, the default option object will be kicked in, so it is still necessary.

File: app/.server/modules/auth/models/user.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+//   merchantId: orNull(bigint()),
+//   phoneNumber: orNull(string()),
+//   createdAt: date(),
+//   updatedAt: orNull(date()),
+//   deletedAt: orNull(date()),
+// });
+
+export const normalizeEmail = (s: string) => s.toLowerCase();
+
+export const normalizeUsername = (s: string) => s.replace(/[^a-zA-Z0-9_]/g, '_').toLowerCase();
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/models/user.ts:30]: the name of the util function implies that its job is not validating email, it just lowercase possibly uppercased letters in email to ensure they can be matched in the future when stored in database or elsewhere.

File: app/.server/api/middlewares.test.ts [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+  'eyJhbGciOiJIUzUxMiJ9.eyJwYXlsb2FkIjp7InNlc3Npb25JZCI6IjM5IiwidXNlciI6eyJpZCI6IjIxIiwiZW1haWwiOiJleGFtcGxlQGdtYWlsLmNvbSJ9fSwidHlwZSI6ImFjY2Vzc190b2tlbiIsImF6cCI6Imh0dHA6Ly9sb2NhbGhvc3Q6NTE3MyIsImlhdCI6MTc0MTM1ODgxMSwiZXhwIjo0ODY1NTYxMjExfQ.jZ91vx6eX6wxMg6sCsbJvKlY4DoYoQJ5Vk26OIwYPBgunInBVhyjkwqSvFCyz_e5Vxm4vn_N3HcVpOpRopa-OA';
+// const validRefreshToken = 'eyJhbGciOiJIUzUxMiJ9.eyJwYXlsb2FkIjp7InNlc3Npb25JZCI6IjM5IiwidXNlciI6eyJpZCI6IjIxIiwiZW1haWwiOiJleGFtcGxlQGdtYWlsLmNvbSJ9fSwidHlwZSI6InJlZnJlc2hfdG9rZW4iLCJhenAiOiJodHRwOi8vbG9jYWxob3N0OjUxNzMiLCJpYXQiOjE3NDEzNTg4NTgsImV4cCI6NDg2NTU2MTI1OH0.VLSnEvdTWscsmhhFUgP1XLVKEF7rpvJsXbMVrB_PEQQDVO71VfLcyNeFZhJIZv979oSKQENurSz2LBtLuceM8g';
+
+describe('getRequestData()', () => {
+  it('should return token and tokenPayload', async () => {
+    // implement tests for getRequestData() at `./middlewares.ts`
+    const { token, tokenPayload } = await getRequestData({
+      request: new Request('https://example.com', {
+        headers: new Headers({
+          Authorization: 'Bearer ' + validAccessToken,
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/middlewares.test.ts:176]: not necessary, concatenating strings is better performance wise

File: app/.server/api/middlewares.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    requestBody = await parseFormData(request, parseHandler);
+  }
+  const validationMessage = await predicate(requestBody);
+  if (validationMessage) {
+    throw getBadRequestResponse({
+      code: apiErrorCodes.INVALID_REQUEST_BODY,
+      message: validationMessage,
+    });
+  }
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/middlewares.ts]: how to cache data when chaining middlewares for one request in RemixJS (remix.run) framework?

File: app/.server/modules/auth/models/user.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  const { [getTableName(usersInHf)]: user, [getTableName(userCredentialsInHf)]: userCredential } = rows[0];
+  if (!userCredential?.password) return null;
+
+  const { passwd } = await getSaltedPassword(password, userCredential.salt);
+  const isValid = await comparePassword(passwd, userCredential.password);
+  if (!isValid) return null;
+
+  return user;
+};
+
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/models/user.ts]: can you be specific on how to do that to ensure best security?

File: app/.server/modules/auth/models/user.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+  const [userCredentialsRes] = await dbInstance
+    .insert(userCredentialsInHf)
+    .values({
+      ...pick(user, ['password', 'salt', 'email']),
+      userId: userRes.id,
+    })
+    .returning();
+
+  return { ...userRes, email: userCredentialsRes.email };
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/models/user.ts]: For real-world app performance, is checking existing user first instead of just insert then handle error from DB is better?

File: app/utils/data.test.ts [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+  it('should work correctly', () => {
+    expect(omit({ b: 1, a: 2 }, ['a'])).toStrictEqual({ b: 1 });
+  });
+
+  // TODO more tests for edge cases
+});
+
+describe('uniq helpers', () => {
+  let origArr: any[], uniqArr: any[];
+  beforeEach(() => {
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/utils/data.test.ts:211]: not necessary for declaring just only 2 variables in one `let` statement as it is still readable enough for human. Alert me when there is more than 2 variables declared in one statement.

File: app/.server/modules/auth/index.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  invariant(import.meta.env.PUBLIC_ORIGIN, 'Missing env variable `PUBLIC_ORIGIN`');
+  return signJwt(
+    {
+      payload,
+      type: 'access_token',
+      azp: import.meta.env.PUBLIC_ORIGIN,
+    },
+    { expirationTime: expirationTime || '1d' },
+  );
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: what is common period to set expiration time for access token & refresh token following standards & ensuring best security?

File: app/.server/modules/auth/index.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  invariant(import.meta.env.PUBLIC_ORIGIN, 'Missing env variable `PUBLIC_ORIGIN`');
+  return signJwt(
+    {
+      payload,
+      type: 'refresh_token',
+      azp: import.meta.env.PUBLIC_ORIGIN,
+    },
+    { expirationTime: expirationTime || '30d' },
+  );
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: what is common period to set expiration time for refresh token following standards & ensuring best security?

File: app/.server/modules/auth/index.ts [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  // Validate the token's authorized party (azp) claim:
+  invariant(import.meta.env.PUBLIC_ORIGIN, 'Missing env variable `PUBLIC_ORIGIN`');
+  const permittedOrigins = [import.meta.env.PUBLIC_ORIGIN];
+  if (typeof payload.azp === 'string' && !permittedOrigins.includes(payload.azp)) {
+    throw new Error('Invalid `azp` claim');
+  }
+  return payload;
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: Thanks. This will be updated in later PRs.

File: app/.server/modules/auth/models/session.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    dbInstance: db,
+  },
+) =>
+  // `SELECT * FROM ${authSessionsInHf} WHERE ${authSessionsInHf.id} = ${id}`
+  (
+    await dbInstance
+      .selectDistinct()
+      .from(authSessionsInHf)
+      .where(and(eq(authSessionsInHf.id, BigInt(sessionId)), gt(authSessionsInHf.expiredAt, new Date())))
+  )[0];
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/models/session.ts]: does using `select distinct` SQL query is slower than normal `select` query? if not, I would like to keep it that way to be more explicit about my intention.

File: app/.server/api/auth/index.ts [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      await signUserOut(refreshToken); // If this fails, we still need to delete the session/tokens from the client and it doesn't do any harm staying in the db anyway.
+    } catch (err) {
+      if (err instanceof Response) throw err;
+      console.error(err);
+      return getGeneralServerErrorResponse();
+    }
+
+    return new Response(null, { status: 204 });
+  }
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/auth/index.ts]: this is my intention to let Remix server take care of "not found" response for unhandled routes for me.

File: app/.server/api/middlewares.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/middlewares.ts]: Nice. Do I need to clean up this cache at the end of each request with WeakMap implementation? how to do it properly for better maintainability?

File: app/.server/utils/jwt.test.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+  });
+
+  it('should throw when payload is invalid', async () => {
+    // @ts-expect-error Testing invalid argument
+    expect(() => signJwt('undefined')).toThrow();
+    // @ts-expect-error Testing invalid argument
+    expect(() => signJwt(undefined)).toThrow();
+    // @ts-expect-error Testing invalid argument
+    expect(() => signJwt(null)).toThrow();
+  });
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/utils/jwt.test.ts:73]: Is it good practice to handle error by its message? I don't think it is due to error message can be changed sometimes and unit tests should not test implementation details too much to avoid time consuming when refactoring.

File: app/.server/api/middlewares.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/middlewares.ts:130]: please explain why do you have global `requestDataCache` and function scoped `cache` inside `getRequestCache` function? why don't just use the global `requestDataCache` instead the function-scoped `cache`?

File: app/.server/modules/auth/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: I want to know how to implement all 3 security recommendations you mentioned, especially token rotation (invalidate refresh token after use) and revocation mechanism.

File: app/.server/modules/auth/models/user.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/models/user.ts]: I want to know more about `INSERT ... ON CONFLICT` SQL statement and how it can help improving performance & code base quality in this case.

File: app/.server/modules/auth/index.ts [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    };
+  },
+>({
+  tokenPayload,
+}: {
+  tokenPayload: T | null;
+}): Promise<boolean> => {
+  const sessionId = tokenPayload?.payload?.sessionId;
+  return !!sessionId && (await getAuthSessionById(sessionId))?.userId?.toString() === tokenPayload.payload?.user.id;
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: please give me full code refactoring this function to do what you suggest, remember to double check the possible issues & bugs, ensure best code quality, maintainability & security.

File: app/services/api.ts [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+  INVALID_AUTH_TOKEN_TYPE: 'invalid_auth_token_type',
+  INVALID_AUTH_CODE: 'invalid_auth_code',
+  INVALID_USERNAME: 'invalid_username',
+  INVALID_PASSWORD: 'invalid_password',
+  INVALID_EMAIL: 'invalid_email',
+  USERNAME_ALREADY_EXISTS: 'username_already_exists',
+  EMAIL_ALREADY_EXISTS: 'email_already_exists',
+  TOO_MANY_REQUESTS: 'too_many_requests',
+  INVALID_REQUEST_METHOD: 'invalid_request_method',
+} as const;
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/services/api.ts:169]: wrong suggestion

File: app/.server/modules/auth/index.ts [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+  // const authSession = await authSessionStorage.getSession(request.headers.get('cookie'));
+  // const sessionId = authSession.get(sessionIdKey);
+  const tokenPayload = await verifyJwt<RefreshTokenPayload>(refreshToken);
+  const sessionId = tokenPayload.payload.payload.sessionId;
+  const userId = tokenPayload.payload.payload.user.id;
+
+  if (sessionId) {
+    void (await deleteAuthSessionById(sessionId));
+  } else console.error('No auth session ID found. User ID:', userId);
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: you mentioned:
> "...as a missing session ID can indicate tampered tokens or system inconsistencies"

Can you explain me this?

File: app/.server/modules/auth/index.ts [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+  invariant(import.meta.env.PUBLIC_ORIGIN, 'Missing env variable `PUBLIC_ORIGIN`');
+  return signJwt(
+    {
+      payload,
+      type: 'refresh_token',
+      azp: import.meta.env.PUBLIC_ORIGIN,
+    },
+    { expirationTime: expirationTime || '30d' },
+  );
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: already done, please recheck the other helper which generates access token in this file.

File: app/.server/api/middlewares.ts [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+
+    const validationMessage = await predicate(requestBody);
+    if (validationMessage) {
+      throw getBadRequestResponse({
+        code: apiErrorCodes.INVALID_REQUEST_BODY,
+        message: validationMessage,
+      });
+    }
+  };
+
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/api/middlewares.ts]: what do you mean by "Unbounded JSON parsing might leave the server exposed to memory exhaustion"?

File: app/.server/utils/error.ts [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+export class AuthError extends ServerBaseError {}
+
+export class ValidationError extends ServerBaseError {
+  constructor(args: PartialBy<ConstructorParameters<typeof ServerBaseError>[0], 'code'>) {
+    super({
+      ...args,
+      code: apiErrorCodes.UNKNOWN_ERROR,
+    });
+  }
+}
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/utils/error.ts:47]: `code` property of `ValidationError` is used for public error message and is overwritable whenever we need to show user the explaination of error.

File: app/.server/utils/jwt.ts [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+  }: JWTVerifyOptions & {
+    secret?: KeyLike | Uint8Array;
+  } = {},
+): Promise<JWTVerifyResult<T>> => {
+  if (!secret) invariant(process.env.JWT_SECRET, 'Missing env variable `JWT_SECRET`');
+  return jwtVerify<T>(token, secret || Buffer.from(process.env.JWT_SECRET), {
+    algorithms: ['HS512', 'RS512'],
+    ...options,
+  });
+};
```
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/utils/jwt.ts]: RS512 is considerred to be used in the future so it is there.
What do you mean "potential security pitfalls". What are they?

File: app/.server/modules/auth/index.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (lvnam96) [reply app/.server/modules/auth/index.ts]: Please checkout `createPayloadForNewTokens` method in `app/.server/modules/auth/index.ts` file in the codebase for unified token payload structure generation and tell me if it is enough. If not, please suggest specific improvements like you said.
